### PR TITLE
Conservative tracer equation in 2D

### PR DESCRIPTION
**Pull request by stephankramer**
Conservative tracer equation, guarantees conservation of tracer also with wetting and drying and discontinuous bathymetry. New implementation as discussed, now solves for depth-integrated `q=Hc` tracer. Includes tests. Separated depth functions (based on wetting and drying options and nonlinear switch) in separate class to make these available in both `ShallowWaterTerm`s and `TracerTerm`s but also in diagnostics.

### CODE REVIEW

File: test/tracerEq/test_steady_adv-diff_mms_2d.py [hunk 1]
```diff
@@ -198,20 +258,36 @@ def timestepper_type(request):
...
-def test_convergence(setup, timestepper_type, auto_sipg):
+def test_convergence(setup, timestepper_type, conservative_form):
     run_convergence(setup, [1, 2, 3], save_plot=False, timestepper_type=timestepper_type,
-                    use_automatic_sipg_parameter=auto_sipg)
+                    use_automatic_sipg_parameter=False,
+                    use_tracer_conservative_form=conservative_form)
+
+
+# only setup2 has nonzero diffusion:
+def test_consergence_auto_sipg(timestepper_type, conservative_form):
```
Comment (tkarna) [inline test/tracerEq/test_steady_adv-diff_mms_2d.py:280]: typo: convergence

File: test/tracerEq/test_steady_adv-diff_mms_2d.py [hunk 2]
```diff
@@ -198,20 +258,36 @@ def timestepper_type(request):
...
+    run_convergence(Setup2, [1, 2, 3], save_plot=False, timestepper_type=timestepper_type,
+                    use_automatic_sipg_parameter=True,
+                    use_tracer_conservative_form=conservative_form)
+
+
+# setup4 is valid for conservative only
+def test_consergence_conservative_only(timestepper_type):
```
Comment (tkarna) [inline test/tracerEq/test_steady_adv-diff_mms_2d.py:287]: typo: convergence

File: thetis/shallowwater_eq.py [hunk 3]
```diff
@@ -208,13 +208,12 @@ class ShallowWaterTerm(Term):
     """
     Generic term in the shallow water equations that provides commonly used
     members and mapping for boundary functions.
-    """
+        """
```
Comment (tkarna) [inline thetis/shallowwater_eq.py:211]: bad indentation?

File: thetis/shallowwater_eq.py [hunk 4]
```diff
@@ -856,25 +829,21 @@ class ShallowWaterEquations(BaseShallowWaterEquation):
...
         u_space, eta_space = function_space.split()
 
-        self.add_momentum_terms(u_test, u_space, eta_space,
-                                bathymetry, options)
+        self.add_momentum_terms(u_test, u_space, eta_space, depth, options)
 
-        self.add_continuity_terms(eta_test, eta_space, u_space, bathymetry, options)
-        self.bathymetry_displacement_mass_term = BathymetryDisplacementMassTerm(eta_test, eta_space, u_space, bathymetry, options)
+        self.add_continuity_terms(eta_test, eta_space, u_space, depth, options)
+        self.bathymetry_displacement_mass_term = BathymetryDisplacementMassTerm(eta_test, eta_space, u_space, depth, options)
```
Comment (tkarna) [inline thetis/shallowwater_eq.py:846]: long line

File: thetis/shallowwater_eq.py [hunk 5]
```diff
@@ -937,21 +901,18 @@ class FreeSurfaceEquation(BaseShallowWaterEquation):
...
-        :type bathymetry: :class:`Function` or :class:`Constant`
+        :arg depth: :class: `DepthExpression` containing depth info
         :arg options: :class:`.AttrDict` object containing all circulation model options
         """
-        super(FreeSurfaceEquation, self).__init__(eta_space, bathymetry, options)
-        self.add_continuity_terms(eta_test, eta_space, u_space, bathymetry, options)
-        self.bathymetry_displacement_mass_term = BathymetryDisplacementMassTerm(eta_test, eta_space, u_space, bathymetry, options)
+        super(FreeSurfaceEquation, self).__init__(eta_space, depth, options)
+        self.add_continuity_terms(eta_test, eta_space, u_space, depth, options)
+        self.bathymetry_displacement_mass_term = BathymetryDisplacementMassTerm(eta_test, eta_space, u_space, depth, options)
```
Comment (tkarna) [inline thetis/shallowwater_eq.py:915]: long line

File: thetis/solver2d.py [hunk 6]
```diff
@@ -289,20 +290,29 @@ def create_equations(self):
...
         )
         self.eq_sw.bnd_functions = self.bnd_functions['shallow_water']
         if self.options.solve_tracer:
             self.fields.tracer_2d = Function(self.function_spaces.Q_2d, name='tracer_2d')
             if self.options.timestepper_type == 'CrankNicolson':
-                self.eq_tracer = tracer_eq_2d.TracerEquation2D(self.function_spaces.Q_2d, bathymetry=self.fields.bathymetry_2d,
-                                                               use_lax_friedrichs=self.options.use_lax_friedrichs_tracer,
-                                                               sipg_parameter=self.options.sipg_parameter_tracer)
+                if self.options.use_tracer_conservative_form:
+                    self.eq_tracer = conservative_tracer_eq_2d.ConservativeTracerEquation2D(self.function_spaces.Q_2d, self.depth,
```
Comment (tkarna) [inline thetis/solver2d.py:309]: long line

File: thetis/utility.py [hunk 7]
```diff
@@ -1726,3 +1724,41 @@ def move_to_nearest_cell_center(location):
         return accepted_locations
     else:
         return accepted_locations, accepted_names
+
+
+class DepthExpression:
+    """
+    Construct expression for depth depending on options
```
Comment (tkarna) [inline thetis/utility.py:1731]: Could use a more descriptive docstring?

File: thetis/shallowwater_eq.py (hunk 4 above)
Comment (stephankramer) [reply thetis/shallowwater_eq.py:846]: shorter than before ;-)
fixed though

File: thetis/utility.py [hunk 8]
```diff
@@ -1727,8 +1727,30 @@ def move_to_nearest_cell_center(location):
...
+
+    If `not use_nonlinear_equations`, then the depth is simply the bathymetry:
+        :math:`H = h`
+    Otherwise we include the free surface elevation:
+        :math:`H = h + \eta`
+    and if `use_wetting_and_drying`, includes a bathymetry displacement term
+    to ensure a positive depth (see Karna et al. 2011):
+        :math:`H = h + f(h+\eta) + \eta`
+    where
+        :math:`f(h+\eta) = (\sqrt{(h+\eta)^2 +\alpha^2} - (h+\eta))/2
```
Comment (tkarna) [inline thetis/utility.py:1742]: missing closing backtick

File: thetis/utility.py [hunk 9]
```diff
@@ -1727,8 +1727,30 @@ def move_to_nearest_cell_center(location):
...
+    This introduces a wetting-drying parameter :math:`\alpha`, with dimensions
+    of length. The value for :math:`\alpha` is specified by
+    `wetting_and_drying_alpha`, in units of meters. The default value is 0.5,
```
Comment (tkarna) [inline thetis/utility.py:1745]: This should generate the correct link
```
:attr:`.ModelOptions.wetting_and_drying_alpha`
```

File: thetis/utility.py [hunk 10]
```diff
@@ -1727,8 +1727,30 @@ def move_to_nearest_cell_center(location):
...
+    but the appropriate value is problem specific and should be set by the user.
+
+    An approximate method for selecting a suitable value for :math:`\alpha` is suggested
```
Comment (tkarna) [inline thetis/utility.py:1748]: This bit is already in the doc of `shallowwater_eq.py`. Maybe no need to duplicate here?

**Pull request merged**